### DISCUSSION TITLE
Details about nonlinearity calculation

### DISCUSSION THREAD

**Discussion by iamfedoroff** [NanoComp/meep#2600] [Q&A] [answered]
Hello,

In the original Meep [paper](https://doi.org/10.1016/j.cpc.2009.11.008) I see that in case of nonlinear effects of the form (see Eq. (3))
$$\tilde{D}_i = \varepsilon^{(1)}E_i + \chi^{(2)} E^2_i + \chi^{(3)}|\mathbf{E}|^2E_i,$$
Meep uses the following update equation for $E_i$ which is based on a Padé approximant (see Eq. (4) and the corresponding [place](https://github.com/NanoComp/meep/blob/90bc78e59affd916d307e32c545f03a79df00f0c/src/step_generic.cpp#L342) in the code):
$$E_i = \left[\frac{1 + \frac{\chi^{(2)}}{[\varepsilon^{(1)}]^2} \tilde{D}_i + 2\frac{\chi^{(3)}}{[\varepsilon^{(1)}]^3} \|| \mathbf{\tilde{D}}\||^2}{1 + 2\frac{\chi^{(2)}}{[\varepsilon^{(1)}]^2} \tilde{D}_i + 3\frac{\chi^{(3)}}{[\varepsilon^{(1)}]^3} \|| \mathbf{\tilde{D}}\||^2}\right] \frac{\tilde{D}_i}{\varepsilon^{(1)}}$$

Can you please explain me how did you obtain this formula.

I would expect that using the above Taylor series for $\tilde{D}_i$ you rewrite it with some Padé approximant, for example
$$\tilde{D}_i(E_i) = \frac{a_0 + a_1E_i + a_2E_i^2}{1 + b_1E_i} + O(E_i^4),$$
and then somehow take its inverse $E_i(D_i)$. But it seems that this is not the case. Can you please help me.

**Comment by stevengj** [comment]
I seem to recall using Maple to solve for the coefficients of the Pade approximant from the Taylor expansion for D.

**Reply by iamfedoroff** [reply]
Well, ok. But what is the logic behind the above formula?
The formula looks like the Padé approximant for some $E_i(\tilde{D_i})$ expression:
$$E_i(\tilde{D_i}) = \frac{a_0 + a_1\tilde{D_i} + a_2\tilde{D_i}^2 + a_3\tilde{D_i}^3}{1 + b_1 \tilde{D_i} + b_2 \tilde{D_i}^2 + b_3\tilde{D_i}^3} + O(\tilde{D_i}^7).$$
If I understand correctly what the Padé approximants are, there should be some Taylor expansion for $E_i(\tilde{D_i})$ with at least six terms:
$$E_i(\tilde{D_i}) = c_0 + c_1\tilde{D_i} + c_2\tilde{D_i}^2 + c_3\tilde{D_i}^3 + c_4\tilde{D_i}^4 + c_5\tilde{D_i}^5 + c_6\tilde{D_i}^6 +  O(\tilde{D_i}^7),$$
where the $c_n$ coefficients define the $a_n$ and $b_n$ coefficients of the Padé approximant for $E_i(\tilde{D_i})$. But how did you obtain this Taylor expansion having only the original cubic formula for $\tilde{D_i}(E_i)$?

**Reply by stevengj** [reply, marked as answer]
There is a cubic equation for the roots of a cubic, and you can Taylor expand from that using first-year-calculus techniques, in principle.      There are also various ways to get an [inverse Taylor series](https://randorithms.com/2021/08/31/Taylor-Series-Inverse.html) without explicitly inverting a function.  Computer-algebra systems like Maple or Mathematica can do all this for you automatically.

I think I just used the Maple [`pade` function](https://www.maplesoft.com/support/help/maple/view.aspx?path=numapprox%2Fpade), if I recall correctly.

**Reply by iamfedoroff** [reply]
Great! Thank you.

**Comment by iamfedoroff** [comment]
Hello! Does anybody know/remember how the above formula was obtained? I do not need a full derivation, just a couple of intermediate steps in order to understand where to dig.
